### PR TITLE
Surface pod provisioning errors

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

The pod spec and/or config such as securityContext may not allow any workload pods to be provisioned by the statefulset or deployment controller. These errors are now surfaced in juju status.

## QA steps

bootstrap k8s and deploy a charm with a bad pod spec
check juju status shows the error

## Bug reference

https://bugs.launchpad.net/juju/+bug/1814055

### CODE REVIEW

File: apiserver/params/params.go [hunk 1]
```diff
@@ -375,6 +375,9 @@ type UpdateApplicationUnitArgs struct {
 type UpdateApplicationUnits struct {
 	ApplicationTag string                  `json:"application-tag"`
 	Scale          *int                    `json:"scale,omitempty"`
+	Status         string                  `json:"status,omitempty"`
```
Comment (ycliuhw) [inline apiserver/params/params.go]: do we wanna have a status struct field instead here?

File: worker/caasunitprovisioner/application_worker.go [hunk 2]
```diff
@@ -169,15 +171,19 @@ func (aw *applicationWorker) loop() error {
...
 			}
-			if scale == nil || *scale == lastReportedScale {
+			var scale *int
+			if service != nil {
+				scale = service.Scale
+			}
+			if scale != nil && *scale == lastReportedScale {
 				continue
 			}
 			lastReportedScale = *scale
```
Comment (ycliuhw) [inline worker/caasunitprovisioner/application_worker.go]: is it possible `scale == nil` here?

File: apiserver/params/params.go (hunk 1 above)
Comment (wallyworld) [reply apiserver/params/params.go]: Yeah, I was being consistent with other places, but will change.

File: worker/caasunitprovisioner/application_worker.go (hunk 2 above)
Comment (wallyworld) [reply worker/caasunitprovisioner/application_worker.go]: Ah yes it is. We (now) need to continue to send status even if scale is nil but as you point out, I need to change the lastReportedScale logic.

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
[caas models - better surfacing of pod errors](https://trello.com/c/YspDN1ok/689-caas-models-better-surfacing-of-pod-errors)

**Pull request merged**